File: C/src/bruteforce.c
#include <stdbool.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include "bruteforce.h"

long bruteforce_randomly(char password[], int min_length, int max_length)
{
    srand(time(NULL));
    const char ALPHANUMERIC[62] = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz1234567890";
    int alphanumeric_length = sizeof(ALPHANUMERIC) - 1;
    bool is_not_cracked = true;
    unsigned long guesses = 1;

    // Allocate max length variable to be reused.
    char *current_guess = malloc(sizeof(char) * (max_length + 1));

    while (is_not_cracked)
    {
        int length = rand() % (max_length - min_length + 1) + min_length;

        for (int i = 0; i < length; i++)
        {
            int random_number = rand() % alphanumeric_length;
            char random_char = ALPHANUMERIC[random_number];
            current_guess[i] = random_char;
        }
        // Terminate leftovers from previous guesses.
        current_guess[length] = '\0'; 

        printf("%lu guess is: %s \n", guesses, current_guess);

        if (strcmp(password, current_guess) == 0)
        {
            printf("Found a match!! \n");
            printf("%s \n", current_guess);
            is_not_cracked = false;
        }

        guesses++;
    }

    // https://www.youtube.com/watch?v=Fgb8l2Ve8vo
    free(current_guess);

    return guesses;
}

File: Rust/src/bruteforce.rs
use std::convert::TryFrom;
use rand::{distributions::Alphanumeric, Rng};

fn get_guess_length(min_length: u32, max_length: u32) -> usize {
    if min_length == max_length {
        return usize::try_from(min_length).unwrap();
    }

    let guess: u32 = rand::thread_rng()
        .gen_range(min_length..max_length+1);
        
    usize::try_from(guess).unwrap()
}

fn make_a_guess(length: usize) -> String {
    rand::thread_rng()
        .sample_iter(&Alphanumeric)
        .take(length)
        .map(char::from)
        .collect()
}

pub fn guess_password(password: String, min_length: u32, max_length: u32) -> i32 {
    let mut guesses: i32 = 1;
    let mut is_not_cracked: bool = true;

    while is_not_cracked {
        let guess_length = get_guess_length(min_length, max_length);
        let guess: String = make_a_guess(guess_length);

        if guess == password {
            is_not_cracked = false;
        }

        guesses += 1;

        println!("{} guess is: {}", guesses, guess);
    }

    guesses
}